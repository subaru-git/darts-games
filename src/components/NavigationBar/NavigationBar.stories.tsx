import type { Meta, StoryObj } from '@storybook/react';
import NavigationBar from './NavigationBar';

const meta: Meta<typeof NavigationBar> = {
  component: NavigationBar,
  title: 'Components/NavigationBar',
};

export default meta;
type Story = StoryObj<typeof NavigationBar>;
export const Default: Story = {
  args: {
    items: [
      {
        label: 'Games',
        children: [
          {
            label: '501',
            subLabel: 'The popular dart game',
            href: '#',
          },
          {
            label: "Eagle's Eye",
            subLabel: 'A dart game for BULL practice',
            href: '#',
          },
          {
            label: 'Cricket Number Count',
            subLabel: 'A original dart game for practice. designed by kikuyama.',
            href: '#',
          },
        ],
      },
      {
        label: 'Respect',
        href: '#',
      },
    ],
  },
};
