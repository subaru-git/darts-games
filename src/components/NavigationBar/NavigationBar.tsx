import React, { FC } from 'react';
import {
  Box,
  Flex,
  Text,
  IconButton,
  Stack,
  Collapse,
  Icon,
  Popover,
  PopoverTrigger,
  PopoverContent,
  useColorModeValue,
  useDisclosure,
  Image,
  Spacer,
  Heading,
} from '@chakra-ui/react';
import Link from 'next/link';
import { GiHamburgerMenu } from 'react-icons/gi';
import { GoChevronDown, GoChevronRight } from 'react-icons/go';
import { GrClose } from 'react-icons/gr';
import LanguageChangeButton from '@/components/LanguageChangeButton';
import useLocale from '@/hooks/locale';

const NavigationBar: FC = () => {
  const { isOpen, onToggle } = useDisclosure();
  const items = GetNavItem();

  return (
    <Box>
      <Flex
        bg={useColorModeValue('white', 'gray.800')}
        color={useColorModeValue('gray.600', 'white')}
        minH={'60px'}
        py={{ base: 2 }}
        px={{ base: 4 }}
        borderBottom={1}
        borderStyle={'solid'}
        borderColor={useColorModeValue('gray.200', 'gray.900')}
        align={'center'}
      >
        <Flex
          flex={{ base: 1, md: 'auto' }}
          ml={{ base: -2 }}
          display={{ base: 'flex', md: 'none' }}
        >
          <IconButton
            onClick={onToggle}
            icon={
              isOpen ? <GrClose width={3} height={3} /> : <GiHamburgerMenu width={5} height={5} />
            }
            variant={'ghost'}
            aria-label={'Toggle Navigation'}
          />
        </Flex>
        <Flex flex={{ base: 'auto' }} justify={{ base: 'center', md: 'start' }}>
          <Box
            _hover={{
              textDecoration: 'none',
              color: useColorModeValue('gray.800', 'white'),
            }}
          >
            <Link href='/' passHref>
              <a>
                <Heading as='h1'>
                  <Image src='/logo.svg' alt='darts homes logo' h='40px' w='200px' />
                </Heading>
              </a>
            </Link>
          </Box>
          <Flex display={{ base: 'none', md: 'flex' }} ml={10}>
            <DesktopNav items={items} />
          </Flex>
          <Spacer />
          <LanguageChangeButton />
        </Flex>
      </Flex>
      <Collapse in={isOpen} animateOpacity>
        <MobileNav items={items} />
      </Collapse>
    </Box>
  );
};

const DesktopNav: FC<{ items: NavItem[] }> = ({ items }) => {
  const linkColor = useColorModeValue('gray.600', 'gray.200');
  const linkHoverColor = useColorModeValue('gray.800', 'white');
  const popoverContentBgColor = useColorModeValue('white', 'gray.800');

  return (
    <Stack direction={'row'} spacing={4} alignItems='center'>
      {items.map((navItem) => (
        <Box key={navItem.label}>
          <Popover trigger={'hover'} placement={'bottom-start'}>
            <PopoverTrigger>
              <Box
                p={2}
                fontSize={'sm'}
                fontWeight={500}
                color={linkColor}
                _hover={{
                  textDecoration: 'none',
                  color: linkHoverColor,
                }}
              >
                <Link href={navItem.href ?? '#'} passHref aria-label={navItem.label}>
                  {navItem.label}
                </Link>
              </Box>
            </PopoverTrigger>

            {navItem.children && (
              <PopoverContent
                border={0}
                boxShadow={'xl'}
                bg={popoverContentBgColor}
                p={4}
                rounded={'xl'}
                minW={'sm'}
              >
                <Stack maxH='80vh' overflowY='scroll'>
                  {navItem.children.map((child) => (
                    <DesktopSubNav key={child.label} {...child} />
                  ))}
                </Stack>
              </PopoverContent>
            )}
          </Popover>
        </Box>
      ))}
    </Stack>
  );
};

const DesktopSubNav = ({ label, href, subLabel }: NavItem) => {
  return (
    <Box
      role={'group'}
      display={'block'}
      p={2}
      rounded={'md'}
      _hover={{ bg: useColorModeValue('pink.50', 'gray.900') }}
    >
      <Link href={href ?? '/'}>
        <a aria-label={label}>
          <Stack direction={'row'} align={'center'}>
            <Box>
              <Text
                transition={'all .3s ease'}
                _groupHover={{ color: 'pink.400' }}
                fontWeight={500}
              >
                {label}
              </Text>
              <Text fontSize={'sm'}>{subLabel}</Text>
            </Box>
            <Flex
              transition={'all .3s ease'}
              transform={'translateX(-10px)'}
              opacity={0}
              _groupHover={{ opacity: '100%', transform: 'translateX(0)' }}
              justify={'flex-end'}
              align={'center'}
              flex={1}
            >
              <Icon color={'pink.400'} w={5} h={5} as={GoChevronRight} />
            </Flex>
          </Stack>
        </a>
      </Link>
    </Box>
  );
};

const MobileNav: FC<{ items: NavItem[] }> = ({ items }) => {
  return (
    <Stack bg={useColorModeValue('white', 'gray.800')} p={4} display={{ md: 'none' }}>
      {items.map((navItem) => (
        <MobileNavItem key={navItem.label} {...navItem} />
      ))}
    </Stack>
  );
};

const MobileNavItem = ({ label, children, href }: NavItem) => {
  const { isOpen, onToggle } = useDisclosure();

  return (
    <Stack spacing={4} onClick={children && onToggle}>
      <Link href={href ?? '#'} passHref>
        <Flex
          py={2}
          justify={'space-between'}
          align={'center'}
          _hover={{
            textDecoration: 'none',
          }}
        >
          <Text fontWeight={600} color={useColorModeValue('gray.600', 'gray.200')}>
            {label}
          </Text>
          {children && (
            <Icon
              as={GoChevronDown}
              transition={'all .25s ease-in-out'}
              transform={isOpen ? 'rotate(180deg)' : ''}
              w={6}
              h={6}
            />
          )}
        </Flex>
      </Link>

      <Collapse in={isOpen} animateOpacity style={{ marginTop: '0!important' }}>
        <Stack
          mt={2}
          pl={4}
          borderLeft={1}
          borderStyle={'solid'}
          borderColor={useColorModeValue('gray.200', 'gray.700')}
          align={'start'}
        >
          {children &&
            children.map((child) => (
              <Link key={child.label} href={child.href ?? '/'}>
                <Box py={2}>{child.label}</Box>
              </Link>
            ))}
        </Stack>
      </Collapse>
    </Stack>
  );
};

export interface NavItem {
  label: string;
  subLabel?: string;
  children?: Array<NavItem>;
  href?: string;
}

const GetNavItem = () => {
  const { t } = useLocale();

  return [
    {
      label: 'Games',
      children: [
        {
          label: "Eagle's Eye",
          subLabel: t.navigation.eagleseye,
          href: '/eagleseye',
        },
        {
          label: 'Cricket Mark-Up',
          subLabel: t.navigation.cricketmarkup,
          href: '/cricketmarkup',
        },
        {
          label: 'Double Trouble',
          subLabel: t.navigation.doubletrouble,
          href: '/doubletrouble',
        },
        {
          label: 'Sweet 16',
          subLabel: t.navigation.sweet16,
          href: '/sweet16',
        },
        {
          label: 'Tops and Tens',
          subLabel: t.navigation.topsandtens,
          href: '/topsandtens',
        },
        {
          label: 'Two-Dart Combinations',
          subLabel: t.navigation.twodartcombinations,
          href: '/twodartcombinations',
        },
        {
          label: 'Around The Compass',
          subLabel: t.navigation.aroundthecompass,
          href: '/aroundthecompass',
        },
        {
          label: 'Tons Up',
          subLabel: t.navigation.tonsup,
          href: '/tonsup',
        },
        {
          label: 'Route 64',
          subLabel: t.navigation.route64,
          href: '/route64',
        },
        {
          label: 'Eighty Threw',
          subLabel: t.navigation.eightythrew,
          href: '/eightythrew',
        },
        {
          label: 'Shanghai Nights',
          subLabel: t.navigation.shanghainights,
          href: '/shanghainights',
        },
        {
          label: 'Switch Hitter',
          subLabel: t.navigation.switchhitter,
          href: '/switchhitter',
        },
        {
          label: 'Bully Bully',
          subLabel: t.navigation.bullybully,
          href: '/bullybully',
        },
      ],
    },
    {
      label: 'Respect',
      href: '/respect',
    },
    {
      label: 'History',
      href: '/history',
    },
  ];
};
export default NavigationBar;
