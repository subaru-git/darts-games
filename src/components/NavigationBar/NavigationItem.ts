import useLocale from '@/hooks/locale';

export interface NavItem {
  label: string;
  subLabel?: string;
  children?: Array<NavItem>;
  href?: string;
}

export const GetNavItem = () => {
  const { t } = useLocale();
  return [
    {
      label: 'Games',
      href: '/games',
      children: [
        {
          label: "Eagle's Eye",
          subLabel: t.navigation.eagleseye,
          href: '/eagleseye',
        },
        {
          label: 'Cricket Mark-Up',
          subLabel: t.navigation.cricketmarkup,
          href: '/cricketmarkup',
        },
        {
          label: 'Double Trouble',
          subLabel: t.navigation.doubletrouble,
          href: '/doubletrouble',
        },
        {
          label: 'Sweet 16',
          subLabel: t.navigation.sweet16,
          href: '/sweet16',
        },
        {
          label: 'Tops and Tens',
          subLabel: t.navigation.topsandtens,
          href: '/topsandtens',
        },
        {
          label: 'Two-Dart Combinations',
          subLabel: t.navigation.twodartcombinations,
          href: '/twodartcombinations',
        },
        {
          label: 'Around The Compass',
          subLabel: t.navigation.aroundthecompass,
          href: '/aroundthecompass',
        },
        {
          label: 'Tons Up',
          subLabel: t.navigation.tonsup,
          href: '/tonsup',
        },
        {
          label: 'Route 64',
          subLabel: t.navigation.route64,
          href: '/route64',
        },
        {
          label: 'Eighty Threw',
          subLabel: t.navigation.eightythrew,
          href: '/eightythrew',
        },
        {
          label: 'Shanghai Nights',
          subLabel: t.navigation.shanghainights,
          href: '/shanghainights',
        },
        {
          label: 'Switch Hitter',
          subLabel: t.navigation.switchhitter,
          href: '/switchhitter',
        },
        {
          label: 'Bully Bully',
          subLabel: t.navigation.bullybully,
          href: '/bullybully',
        },
        {
          label: 'Treble For Show',
          subLabel: t.navigation.treblesforshow,
          href: '/treblesforshow',
        },
        {
          label: 'Arrange',
          subLabel: t.navigation.arrange,
          href: '/arrange',
        },
        {
          label: 'Count Up',
          subLabel: t.navigation.countup,
          href: '/countup',
        },
      ],
    },
    {
      label: 'Tools',
      children: [
        {
          label: 'Arrange Record',
          subLabel: t.navigation.arrangerecord,
          href: '/arrangerecord',
        },
      ],
    },
    {
      label: 'Respect',
      href: '/respect',
    },
    {
      label: 'History',
      href: '/history',
    },
  ];
};
