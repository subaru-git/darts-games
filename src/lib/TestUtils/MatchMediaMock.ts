import mediaQuery from 'css-mediaquery'

export const createMatchMedia =
  (width: number) =>
  (query: string): MediaQueryList => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })
