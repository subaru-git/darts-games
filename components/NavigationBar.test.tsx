import NavigationBar, { NavItem } from "./NavigationBar";
import { render, screen } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";
import MatchMediaMock from "jest-matchmedia-mock";
let matchMediaMock: MatchMediaMock;
beforeEach(() => (matchMediaMock = new MatchMediaMock()));
afterEach(() => matchMediaMock.clear());

test("should rendering", () => {
  const { container } = render(
    <ChakraProvider>
      <NavigationBar items={items} />
    </ChakraProvider>
  );
  expect(screen.getAllByText("Darts-Games")).toHaveLength(1);
  expect(container).toMatchSnapshot();
});

const items: Array<NavItem> = [
  {
    label: "Games",
    children: [
      {
        label: "501",
        subLabel: "The popular dart game",
        href: "#",
      },
      {
        label: "Eagle's Eye",
        subLabel: "A dart game for BULL practice",
        href: "/eagleseye",
      },
      {
        label: "Cricket Number Count",
        subLabel: "A original dart game for practice. designed by kikuyama.",
        href: "/cricketnumbercount",
      },
    ],
  },
  {
    label: "Respects",
    href: "#",
  },
];
