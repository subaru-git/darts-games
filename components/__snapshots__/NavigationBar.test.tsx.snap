// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should rendering 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-1v5zkfs"
    >
      <div
        class="css-1ibo7k1"
      >
        <button
          aria-label="Toggle Navigation"
          class="chakra-button css-ywwdc4"
          type="button"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M32 96v64h448V96H32zm0 128v64h448v-64H32zm0 128v64h448v-64H32z"
            />
          </svg>
        </button>
      </div>
      <div
        class="css-hhwkx"
      >
        <a
          class="chakra-link css-32dq02"
          href="/"
        >
          <img
            alt="logo"
            class="chakra-image css-xgff9b"
            src="logo.png"
          />
        </a>
        <div
          class="css-s90a0u"
        >
          <div
            class="chakra-stack css-a9v878"
          >
            <div
              class="css-0"
            >
              <a
                aria-controls="popover-content-:r2:"
                aria-expanded="false"
                aria-haspopup="dialog"
                class="chakra-link css-oe8fj1"
                href="#"
                id="popover-trigger-:r2:"
              >
                Games
              </a>
              <div
                class="chakra-popover__popper css-1qq679y"
                style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
              >
                <section
                  class="chakra-popover__content css-24m2y2"
                  id="popover-content-:r2:"
                  role="tooltip"
                  style="transform-origin: var(--popper-transform-origin); opacity: 0; transform: scale(0.95) translateZ(0);"
                  tabindex="-1"
                >
                  <div
                    class="chakra-stack css-n21gh5"
                  >
                    <a
                      class="chakra-link css-1tsr3k3"
                      href="#"
                      role="group"
                    >
                      <div
                        class="chakra-stack css-84zodg"
                      >
                        <div
                          class="css-0"
                        >
                          <p
                            class="chakra-text css-sc37ud"
                          >
                            501
                          </p>
                          <p
                            class="chakra-text css-itvw0n"
                          >
                            The popular dart game
                          </p>
                        </div>
                        <div
                          class="css-2gz105"
                        >
                          <svg
                            class="chakra-icon css-swb7g"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            stroke="currentColor"
                            stroke-width="0"
                            viewBox="0 0 8 16"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M7.5 8l-5 5L1 11.5 4.75 8 1 4.5 2.5 3l5 5z"
                              fill-rule="evenodd"
                            />
                          </svg>
                        </div>
                      </div>
                    </a>
                    <a
                      class="chakra-link css-1tsr3k3"
                      href="/eagleseye"
                      role="group"
                    >
                      <div
                        class="chakra-stack css-84zodg"
                      >
                        <div
                          class="css-0"
                        >
                          <p
                            class="chakra-text css-sc37ud"
                          >
                            Eagle's Eye
                          </p>
                          <p
                            class="chakra-text css-itvw0n"
                          >
                            A dart game for BULL practice
                          </p>
                        </div>
                        <div
                          class="css-2gz105"
                        >
                          <svg
                            class="chakra-icon css-swb7g"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            stroke="currentColor"
                            stroke-width="0"
                            viewBox="0 0 8 16"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M7.5 8l-5 5L1 11.5 4.75 8 1 4.5 2.5 3l5 5z"
                              fill-rule="evenodd"
                            />
                          </svg>
                        </div>
                      </div>
                    </a>
                    <a
                      class="chakra-link css-1tsr3k3"
                      href="/cricketnumbercount"
                      role="group"
                    >
                      <div
                        class="chakra-stack css-84zodg"
                      >
                        <div
                          class="css-0"
                        >
                          <p
                            class="chakra-text css-sc37ud"
                          >
                            Cricket Number Count
                          </p>
                          <p
                            class="chakra-text css-itvw0n"
                          >
                            A original dart game for practice. designed by kikuyama.
                          </p>
                        </div>
                        <div
                          class="css-2gz105"
                        >
                          <svg
                            class="chakra-icon css-swb7g"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            stroke="currentColor"
                            stroke-width="0"
                            viewBox="0 0 8 16"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M7.5 8l-5 5L1 11.5 4.75 8 1 4.5 2.5 3l5 5z"
                              fill-rule="evenodd"
                            />
                          </svg>
                        </div>
                      </div>
                    </a>
                  </div>
                </section>
              </div>
            </div>
            <div
              class="css-0"
            >
              <a
                aria-controls="popover-content-:r4:"
                aria-expanded="false"
                aria-haspopup="dialog"
                class="chakra-link css-oe8fj1"
                href="#"
                id="popover-trigger-:r4:"
              >
                Respects
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="chakra-collapse"
      style="overflow: hidden; display: none; opacity: 0; height: 0px;"
    >
      <div
        class="chakra-stack css-z4006u"
      >
        <div
          class="chakra-stack css-egoftb"
        >
          <a
            class="chakra-link css-wnxme9"
            href="#"
          >
            <p
              class="chakra-text css-9y4vnp"
            >
              Games
            </p>
            <svg
              class="chakra-icon css-q9erop"
              fill="currentColor"
              focusable="false"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 10 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 11L0 6l1.5-1.5L5 8.25 8.5 4.5 10 6l-5 5z"
                fill-rule="evenodd"
              />
            </svg>
          </a>
          <div
            class="chakra-collapse"
            style="overflow: hidden; display: none; opacity: 0; height: 0px;"
          >
            <div
              class="chakra-stack css-11sp92w"
            >
              <a
                class="chakra-link css-km23hw"
                href="#"
              >
                501
              </a>
              <a
                class="chakra-link css-km23hw"
                href="/eagleseye"
              >
                Eagle's Eye
              </a>
              <a
                class="chakra-link css-km23hw"
                href="/cricketnumbercount"
              >
                Cricket Number Count
              </a>
            </div>
          </div>
        </div>
        <div
          class="chakra-stack css-egoftb"
        >
          <a
            class="chakra-link css-wnxme9"
            href="#"
          >
            <p
              class="chakra-text css-9y4vnp"
            >
              Respects
            </p>
          </a>
          <div
            class="chakra-collapse"
            style="overflow: hidden; display: none; opacity: 0; height: 0px;"
          >
            <div
              class="chakra-stack css-11sp92w"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
